{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from Bio import AlignIO\n",
    "import os\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_nearest(pdb, n_nearest):\n",
    "    import numpy as np\n",
    "    from scipy.spatial import distance\n",
    "    \"\"\"\n",
    "    This function finds the n_nearest atoms to a given atom in a pdb file.\n",
    "    \"\"\"\n",
    "    with open(pdb, 'r') as f:\n",
    "        lines = [line for line in f if line.startswith('ATOM') and line[13:15] == 'CA']\n",
    "    coords = np.array([line.split()[6:9] for line in lines], dtype=float)\n",
    "    distance_matrix = distance.cdist(coords, coords)\n",
    "    nearest_ind = np.array([np.argsort(distance_matrix[i])[1:n_nearest+1] for i in range(len(distance_matrix))])\n",
    "    nearest_dist = np.array([np.sort(distance_matrix[i])[1:n_nearest+1] for i in range(len(distance_matrix))])\n",
    "    return nearest_ind, nearest_dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def jsd(p, q):\n",
    "    \"\"\"\n",
    "    This function calculates the Jensen-Shannon divergence between two probability distributions.\n",
    "    \"\"\"\n",
    "    from scipy.special import rel_entr\n",
    "    import numpy as np\n",
    "    m = 0.5 * (p + q)\n",
    "    jsd = 0.5 * np.sum(rel_entr(p, m)) + 0.5 * np.sum(rel_entr(q, m))\n",
    "    return jsd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fasta2dict(fasta_path):\n",
    "    with open(fasta_path) as f:\n",
    "        content = f.read()\n",
    "        entries = content.split('>')[1:]\n",
    "        names = [entry.split('\\n')[0] for entry in entries]\n",
    "        seqs = [entry.split('\\n')[1:] for entry in entries]\n",
    "        seqs = [''.join(seq) for seq in seqs]\n",
    "        dic = dict(zip(names,seqs))\n",
    "        return dic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sum_fasta(seqs:str, headings:str, path:str, other_seq = None, other_heading = None):\n",
    "    import os\n",
    "    if len(seqs) != len(headings):\n",
    "        raise ValueError('Sequence and heading list not of same length')\n",
    "    content = ''\n",
    "    if other_seq != None and other_heading != None:\n",
    "        content += '>'+other_heading+'\\n'+other_seq+'\\n'\n",
    "    for n in range(len(seqs)):\n",
    "        content+='>'+headings[n]+'\\n'\n",
    "        content+=seqs[n]+'\\n'\n",
    "    with open(path,'w') as f:\n",
    "        f.write(content)\n",
    "    return content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mmseqs_cons23(input_fasta : str, database_fasta : str, input_pdb : str = None, windows:bool = False, max_seqs:int = 1000, include_neighbors:bool=False, n_nearest:int=5, decay:float=0.5, alpha:float=0.5, output_path = './data/out_alignment23.fasta'):\n",
    "    import subprocess\n",
    "    import numpy as np\n",
    "    import os\n",
    "    if not os.path.exists('./data/mmseqs_data'):\n",
    "        os.makedirs('./data/mmseqs_data')\n",
    "    subprocess.run([\"mmseqs\",\"createdb\",database_fasta,\"./data/mmseqs_data/database_db23\"])                #create mmseq db file from fasta (database)\n",
    "    subprocess.run([\"mmseqs\",\"createdb\", input_fasta, \"./data/mmseqs_data/protein_db23\"])   \n",
    "    subprocess.run([\"mmseqs\",\"search\",\"./data/mmseqs_data/protein_db23\",\"./data/mmseqs_data/database_db23\",\"./data/mmseqs_data/result23\",\"./data/mmseqs_data/tmp\",\"--max-seqs\",str(max_seqs)])\n",
    "    subprocess.run([\"mmseqs\",\"result2msa\",\"./data/mmseqs_data/protein_db23\",\"./data/mmseqs_data/database_db23\",\"./data/mmseqs_data/result23\",\"./data/mmseqs_data/alignment_msa23.fasta\",\"--msa-format-mode\", \"2\"])     # generate multiple sequence alignment from search result\n",
    "    with open(input_fasta) as f:\n",
    "        text = f.read().split('>')[1]\n",
    "        nameinterest = text.split('\\n')[0]\n",
    "        seqinterest = ''.join(text.split('\\n')[1:])\n",
    "    with open('./data/mmseqs_data/alignment_msa23.fasta_header.ffdata','rb') as f:\n",
    "        text = f.read()\n",
    "        text = text.decode('utf-8').replace('\\x00', '\\t')\n",
    "        prots = text.split('\\t')\n",
    "    prots = [x.split('|')[0] if '|taxID' in x else x for x in prots]\n",
    "    with open('./data/mmseqs_data/alignment_msa23.fasta_sequence.ffdata', 'rb') as f:\n",
    "        text2 = f.read()\n",
    "        text2 = text2.decode('utf-8').replace('\\x00', '\\t')\n",
    "    seqs = text2.split('\\t')\n",
    "    protseq = dict(zip(prots, seqs))\n",
    "    with open('./data/mmseqs_data/alignment23.m8') as f:\n",
    "        content = f.readlines()\n",
    "        alis = [content[x].split('\\t')[1].split('|')[0] for x in range(len(content)) if '|' in content[x].split('\\t')[1].split('|')]\n",
    "    alseqs = [protseq[al].strip() for al in alis]\n",
    "    sum_fasta(alseqs, alis, './data/tobealigned23.fasta', seqinterest, nameinterest)\n",
    "    subprocess.run(['muscle','-align','./data/tobealigned23.fasta','-output','./data/firstalignment23.fasta'])\n",
    "    alignmentdict = fasta2dict('./data/firstalignment23.fasta')\n",
    "    arr = np.array(list(''.join(test[nameinterest])))\n",
    "    inds = np.where(arr != '-')[0]\n",
    "    cleaned = [''.join(np.array(list(seq))[inds]) for seq in list(alignmentdict.values())]\n",
    "    sum_fasta(cleaned, list(temp.keys()), path = output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mmseqs_cons2(input_fasta : str, database_fasta : str, input_pdb : str = None,windows:bool = False, max_seqs:int = 1000, include_neighbors:bool=False, n_nearest:int=5, decay:float=0.5, alpha:float=0.5, output_path = './data/mmseqs_data/alignment_msa.fasta'):\n",
    "    import subprocess\n",
    "    import numpy as np\n",
    "    import os\n",
    "    from Bio import AlignIO\n",
    "    if not os.path.exists('./data/mmseqs_data'):\n",
    "        os.makedirs('./data/mmseqs_data')\n",
    "    subprocess.run([\"mmseqs\",\"createdb\",database_fasta,\"./data/mmseqs_data/database_db\"])                #create mmseq db file from fasta (database)\n",
    "    subprocess.run([\"mmseqs\",\"createdb\", input_fasta, \"./data/mmseqs_data/protein_db\"])   \n",
    "    subprocess.run([\"mmseqs\",\"search\",\"./data/mmseqs_data/protein_db\",\"./data/mmseqs_data/database_db\",\"./data/mmseqs_data/result\",\"./data/mmseqs_data/tmp\",\"--max-seqs\",str(max_seqs)])\n",
    "    subprocess.run([\"mmseqs\",\"result2msa\",\"./data/mmseqs_data/protein_db\",\"./data/mmseqs_data/database_db\",\"./data/mmseqs_data/result\",output_path,\"--msa-format-mode\", \"2\"])     # generate multiple sequence alignment from search result\n",
    "    with open(output_path, \"rb+\") as file:     # remove NUL character in fasta\n",
    "        file.seek(-1, 2)       # move pointer to last byte and go back one more\n",
    "        file.truncate()\n",
    "        \n",
    "    aa_order = ['A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V']\n",
    "    blosum62_background = np.array([0.078, 0.051, 0.041, 0.052, 0.024, 0.034, 0.059, 0.083, 0.025, 0.062, 0.092, 0.056, 0.024, 0.044, 0.043, 0.059, 0.055, 0.014, 0.034, 0.072])\n",
    "    alignment = AlignIO.read('./data/mmseqs_data/alignment_msa.fasta', 'fasta')   # read msa using biopython\n",
    "    alignment_array = np.array([list(rec.seq) for rec in alignment], dtype = 'U1')\n",
    "    aa_prop_array = np.zeros((20, alignment_array.shape[1]))   # initiate array to store amino acid proportions\n",
    "    for n_col, col in enumerate(alignment_array.T):            # calculate amino acid proportions for each column\n",
    "        col_no_gaps = np.char.upper(col[col != '-'])\n",
    "        aa_percs = np.array([np.sum(col_no_gaps == aa)/len(col_no_gaps) for aa in aa_order])    \n",
    "        aa_prop_array[:, n_col] = aa_percs                              \n",
    "    conservation_scores = np.zeros(len(alignment_array.T))\n",
    "    for n_col, col in enumerate(aa_prop_array.T):\n",
    "        conservation_scores[n_col] = jsd(col, blosum62_background)  # compute Jensen-Shannon divergence for each column\n",
    "    if include_neighbors:\n",
    "        ind,dist = find_nearest(input_pdb, n_nearest)\n",
    "        weights = np.array([np.exp(-decay*row) for row in dist])\n",
    "        weightsums = np.sum(weights, axis = 1)\n",
    "        neighbors = conservation_scores[ind]\n",
    "        neighbors_weighted = np.sum(neighbors*weights, axis = 1)/weightsums\n",
    "        conservation_scores_neighbors = alpha * conservation_scores + (1-alpha) * neighbors_weighted\n",
    "        return conservation_scores, conservation_scores_neighbors\n",
    "    else:\n",
    "        return conservation_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "363"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tes = AlignIO.read('./data/mmseqs_data/alignment_msa.fasta','fasta')\n",
    "arr = np.array([list(rec.seq) for rec in tes], dtype = 'U1')\n",
    "len(arr.T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/mmseqs_data/protein_db exists and will be overwritten\n",
      "createdb ./data/Q72KT1.fasta ./data/mmseqs_data/protein_db \n",
      "\n",
      "MMseqs Version:       \t13.45111\n",
      "Database type         \t0\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[\n",
      "Time for merging to protein_db_h: 0h 0m 0s 8ms\n",
      "Time for merging to protein_db: 0h 0m 0s 7ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 31ms\n",
      "./data/mmseqs_data/result exists and will be overwritten\n",
      "search ./data/mmseqs_data/protein_db ./data/mmseqs_data/database_db ./data/mmseqs_data/result ./data/mmseqs_data/tmp --max-seqs 1000 \n",
      "\n",
      "MMseqs Version:                        \t13.45111\n",
      "Substitution matrix                    \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Add backtrace                          \tfalse\n",
      "Alignment mode                         \t2\n",
      "Alignment mode                         \t0\n",
      "Allow wrapped scoring                  \tfalse\n",
      "E-value threshold                      \t0.001\n",
      "Seq. id. threshold                     \t0\n",
      "Min alignment length                   \t0\n",
      "Seq. id. mode                          \t0\n",
      "Alternative alignments                 \t0\n",
      "Coverage threshold                     \t0\n",
      "Coverage mode                          \t0\n",
      "Max sequence length                    \t65535\n",
      "Compositional bias                     \t1\n",
      "Max reject                             \t2147483647\n",
      "Max accept                             \t2147483647\n",
      "Include identical seq. id.             \tfalse\n",
      "Preload mode                           \t0\n",
      "Pseudo count a                         \t1\n",
      "Pseudo count b                         \t1.5\n",
      "Score bias                             \t0\n",
      "Realign hits                           \tfalse\n",
      "Realign score bias                     \t-0.2\n",
      "Realign max seqs                       \t2147483647\n",
      "Gap open cost                          \tnucl:5,aa:11\n",
      "Gap extension cost                     \tnucl:2,aa:1\n",
      "Zdrop                                  \t40\n",
      "Threads                                \t64\n",
      "Compressed                             \t0\n",
      "Verbosity                              \t3\n",
      "Seed substitution matrix               \tnucl:nucleotide.out,aa:VTML80.out\n",
      "Sensitivity                            \t5.7\n",
      "k-mer length                           \t0\n",
      "k-score                                \t2147483647\n",
      "Alphabet size                          \tnucl:5,aa:21\n",
      "Max results per query                  \t1000\n",
      "Split database                         \t0\n",
      "Split mode                             \t2\n",
      "Split memory limit                     \t0\n",
      "Diagonal scoring                       \ttrue\n",
      "Exact k-mer matching                   \t0\n",
      "Mask residues                          \t1\n",
      "Mask lower case residues               \t0\n",
      "Minimum diagonal score                 \t15\n",
      "Spaced k-mers                          \t1\n",
      "Spaced k-mer pattern                   \t\n",
      "Local temporary path                   \t\n",
      "Rescore mode                           \t0\n",
      "Remove hits by seq. id. and coverage   \tfalse\n",
      "Sort results                           \t0\n",
      "Mask profile                           \t1\n",
      "Profile E-value threshold              \t0.1\n",
      "Global sequence weighting              \tfalse\n",
      "Allow deletions                        \tfalse\n",
      "Filter MSA                             \t1\n",
      "Maximum seq. id. threshold             \t0.9\n",
      "Minimum seq. id.                       \t0\n",
      "Minimum score per column               \t-20\n",
      "Minimum coverage                       \t0\n",
      "Select N most diverse seqs             \t1000\n",
      "Min codons in orf                      \t30\n",
      "Max codons in length                   \t32734\n",
      "Max orf gaps                           \t2147483647\n",
      "Contig start mode                      \t2\n",
      "Contig end mode                        \t2\n",
      "Orf start mode                         \t1\n",
      "Forward frames                         \t1,2,3\n",
      "Reverse frames                         \t1,2,3\n",
      "Translation table                      \t1\n",
      "Translate orf                          \t0\n",
      "Use all table starts                   \tfalse\n",
      "Offset of numeric ids                  \t0\n",
      "Create lookup                          \t0\n",
      "Add orf stop                           \tfalse\n",
      "Overlap between sequences              \t0\n",
      "Sequence split mode                    \t1\n",
      "Header split mode                      \t0\n",
      "Chain overlapping alignments           \t0\n",
      "Merge query                            \t1\n",
      "Search type                            \t0\n",
      "Search iterations                      \t1\n",
      "Start sensitivity                      \t4\n",
      "Search steps                           \t1\n",
      "Exhaustive search mode                 \tfalse\n",
      "Filter results during exhaustive search\t0\n",
      "Strand selection                       \t1\n",
      "LCA search mode                        \tfalse\n",
      "Disk space limit                       \t0\n",
      "MPI runner                             \t\n",
      "Force restart with latest tmp          \tfalse\n",
      "Remove temporary files                 \tfalse\n",
      "\n",
      "./data/mmseqs_data/result.dbtype exists already!\n",
      "./data/mmseqs_data/alignment_msa.fasta exists and will be overwritten\n",
      "result2msa ./data/mmseqs_data/protein_db ./data/mmseqs_data/database_db ./data/mmseqs_data/result ./data/mmseqs_data/alignment_msa.fasta --msa-format-mode 2 \n",
      "\n",
      "MMseqs Version:           \t13.45111\n",
      "Substitution matrix       \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Gap open cost             \tnucl:5,aa:11\n",
      "Gap extension cost        \tnucl:2,aa:1\n",
      "Allow deletions           \tfalse\n",
      "Compositional bias        \t1\n",
      "MSA format mode           \t2\n",
      "Summary prefix            \tcl\n",
      "Skip query                \tfalse\n",
      "Filter MSA                \t0\n",
      "Maximum seq. id. threshold\t0.9\n",
      "Minimum seq. id.          \t0\n",
      "Minimum score per column  \t-20\n",
      "Minimum coverage          \t0\n",
      "Select N most diverse seqs\t1000\n",
      "Threads                   \t64\n",
      "Compressed                \t0\n",
      "Verbosity                 \t3\n",
      "\n",
      "Query database size: 1 type: Aminoacid\n",
      "Target database size: 850 type: Aminoacid\n",
      "[=================================================================] 1 0s 0ms\n",
      "Time for merging to alignment_msa.fasta: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 9ms\n"
     ]
    }
   ],
   "source": [
    "from mmseq import mmseqs_cons\n",
    "temp = mmseqs_cons('./data/Q72KT1.fasta','./data/850db.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.63694105, 0.52011183, 0.55386659, 0.46478871, 0.50182057,\n",
       "       0.51919405, 0.51076311, 0.60423363, 0.48755964, 0.46478871])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/mmseqs_data/protein_db exists and will be overwritten\n",
      "createdb ./data/Q72KT1.fasta ./data/mmseqs_data/protein_db \n",
      "\n",
      "MMseqs Version:       \t13.45111\n",
      "Database type         \t0\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[\n",
      "Time for merging to protein_db_h: 0h 0m 0s 7ms\n",
      "Time for merging to protein_db: 0h 0m 0s 7ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 30ms\n",
      "./data/mmseqs_data/result exists and will be overwritten\n",
      "search ./data/mmseqs_data/protein_db ./data/mmseqs_data/database_db ./data/mmseqs_data/result ./data/mmseqs_data/tmp --max-seqs 1000 \n",
      "\n",
      "MMseqs Version:                        \t13.45111\n",
      "Substitution matrix                    \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Add backtrace                          \tfalse\n",
      "Alignment mode                         \t2\n",
      "Alignment mode                         \t0\n",
      "Allow wrapped scoring                  \tfalse\n",
      "E-value threshold                      \t0.001\n",
      "Seq. id. threshold                     \t0\n",
      "Min alignment length                   \t0\n",
      "Seq. id. mode                          \t0\n",
      "Alternative alignments                 \t0\n",
      "Coverage threshold                     \t0\n",
      "Coverage mode                          \t0\n",
      "Max sequence length                    \t65535\n",
      "Compositional bias                     \t1\n",
      "Max reject                             \t2147483647\n",
      "Max accept                             \t2147483647\n",
      "Include identical seq. id.             \tfalse\n",
      "Preload mode                           \t0\n",
      "Pseudo count a                         \t1\n",
      "Pseudo count b                         \t1.5\n",
      "Score bias                             \t0\n",
      "Realign hits                           \tfalse\n",
      "Realign score bias                     \t-0.2\n",
      "Realign max seqs                       \t2147483647\n",
      "Gap open cost                          \tnucl:5,aa:11\n",
      "Gap extension cost                     \tnucl:2,aa:1\n",
      "Zdrop                                  \t40\n",
      "Threads                                \t64\n",
      "Compressed                             \t0\n",
      "Verbosity                              \t3\n",
      "Seed substitution matrix               \tnucl:nucleotide.out,aa:VTML80.out\n",
      "Sensitivity                            \t5.7\n",
      "k-mer length                           \t0\n",
      "k-score                                \t2147483647\n",
      "Alphabet size                          \tnucl:5,aa:21\n",
      "Max results per query                  \t1000\n",
      "Split database                         \t0\n",
      "Split mode                             \t2\n",
      "Split memory limit                     \t0\n",
      "Diagonal scoring                       \ttrue\n",
      "Exact k-mer matching                   \t0\n",
      "Mask residues                          \t1\n",
      "Mask lower case residues               \t0\n",
      "Minimum diagonal score                 \t15\n",
      "Spaced k-mers                          \t1\n",
      "Spaced k-mer pattern                   \t\n",
      "Local temporary path                   \t\n",
      "Rescore mode                           \t0\n",
      "Remove hits by seq. id. and coverage   \tfalse\n",
      "Sort results                           \t0\n",
      "Mask profile                           \t1\n",
      "Profile E-value threshold              \t0.1\n",
      "Global sequence weighting              \tfalse\n",
      "Allow deletions                        \tfalse\n",
      "Filter MSA                             \t1\n",
      "Maximum seq. id. threshold             \t0.9\n",
      "Minimum seq. id.                       \t0\n",
      "Minimum score per column               \t-20\n",
      "Minimum coverage                       \t0\n",
      "Select N most diverse seqs             \t1000\n",
      "Min codons in orf                      \t30\n",
      "Max codons in length                   \t32734\n",
      "Max orf gaps                           \t2147483647\n",
      "Contig start mode                      \t2\n",
      "Contig end mode                        \t2\n",
      "Orf start mode                         \t1\n",
      "Forward frames                         \t1,2,3\n",
      "Reverse frames                         \t1,2,3\n",
      "Translation table                      \t1\n",
      "Translate orf                          \t0\n",
      "Use all table starts                   \tfalse\n",
      "Offset of numeric ids                  \t0\n",
      "Create lookup                          \t0\n",
      "Add orf stop                           \tfalse\n",
      "Overlap between sequences              \t0\n",
      "Sequence split mode                    \t1\n",
      "Header split mode                      \t0\n",
      "Chain overlapping alignments           \t0\n",
      "Merge query                            \t1\n",
      "Search type                            \t0\n",
      "Search iterations                      \t1\n",
      "Start sensitivity                      \t4\n",
      "Search steps                           \t1\n",
      "Exhaustive search mode                 \tfalse\n",
      "Filter results during exhaustive search\t0\n",
      "Strand selection                       \t1\n",
      "LCA search mode                        \tfalse\n",
      "Disk space limit                       \t0\n",
      "MPI runner                             \t\n",
      "Force restart with latest tmp          \tfalse\n",
      "Remove temporary files                 \tfalse\n",
      "\n",
      "./data/mmseqs_data/result.dbtype exists already!\n",
      "./data/mmseqs_data/alignment_msa.fasta exists and will be overwritten\n",
      "result2msa ./data/mmseqs_data/protein_db ./data/mmseqs_data/database_db ./data/mmseqs_data/result ./data/mmseqs_data/alignment_msa.fasta --msa-format-mode 2 \n",
      "\n",
      "MMseqs Version:           \t13.45111\n",
      "Substitution matrix       \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Gap open cost             \tnucl:5,aa:11\n",
      "Gap extension cost        \tnucl:2,aa:1\n",
      "Allow deletions           \tfalse\n",
      "Compositional bias        \t1\n",
      "MSA format mode           \t2\n",
      "Summary prefix            \tcl\n",
      "Skip query                \tfalse\n",
      "Filter MSA                \t0\n",
      "Maximum seq. id. threshold\t0.9\n",
      "Minimum seq. id.          \t0\n",
      "Minimum score per column  \t-20\n",
      "Minimum coverage          \t0\n",
      "Select N most diverse seqs\t1000\n",
      "Threads                   \t64\n",
      "Compressed                \t0\n",
      "Verbosity                 \t3\n",
      "\n",
      "Query database size: 1 type: Aminoacid\n",
      "Target database size: 850 type: Aminoacid\n",
      "[=================================================================] 1 0s 0ms\n",
      "Time for merging to alignment_msa.fasta: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 9ms\n"
     ]
    }
   ],
   "source": [
    "temp_spatial = mmseqs_cons('./data/Q72KT1.fasta','./data/850db.fasta',input_pdb = './data/Q72KT1.pdb', include_neighbors = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/mmseqs_data/protein_db exists and will be overwritten\n",
      "createdb ./data/Q72KT1.fasta ./data/mmseqs_data/protein_db \n",
      "\n",
      "MMseqs Version:       \t13.45111\n",
      "Database type         \t0\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[\n",
      "Time for merging to protein_db_h: 0h 0m 0s 7ms\n",
      "Time for merging to protein_db: 0h 0m 0s 6ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 29ms\n",
      "./data/mmseqs_data/result exists and will be overwritten\n",
      "search ./data/mmseqs_data/protein_db ./data/mmseqs_data/database_db ./data/mmseqs_data/result ./data/mmseqs_data/tmp --max-seqs 1000 \n",
      "\n",
      "MMseqs Version:                        \t13.45111\n",
      "Substitution matrix                    \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Add backtrace                          \tfalse\n",
      "Alignment mode                         \t2\n",
      "Alignment mode                         \t0\n",
      "Allow wrapped scoring                  \tfalse\n",
      "E-value threshold                      \t0.001\n",
      "Seq. id. threshold                     \t0\n",
      "Min alignment length                   \t0\n",
      "Seq. id. mode                          \t0\n",
      "Alternative alignments                 \t0\n",
      "Coverage threshold                     \t0\n",
      "Coverage mode                          \t0\n",
      "Max sequence length                    \t65535\n",
      "Compositional bias                     \t1\n",
      "Max reject                             \t2147483647\n",
      "Max accept                             \t2147483647\n",
      "Include identical seq. id.             \tfalse\n",
      "Preload mode                           \t0\n",
      "Pseudo count a                         \t1\n",
      "Pseudo count b                         \t1.5\n",
      "Score bias                             \t0\n",
      "Realign hits                           \tfalse\n",
      "Realign score bias                     \t-0.2\n",
      "Realign max seqs                       \t2147483647\n",
      "Gap open cost                          \tnucl:5,aa:11\n",
      "Gap extension cost                     \tnucl:2,aa:1\n",
      "Zdrop                                  \t40\n",
      "Threads                                \t64\n",
      "Compressed                             \t0\n",
      "Verbosity                              \t3\n",
      "Seed substitution matrix               \tnucl:nucleotide.out,aa:VTML80.out\n",
      "Sensitivity                            \t5.7\n",
      "k-mer length                           \t0\n",
      "k-score                                \t2147483647\n",
      "Alphabet size                          \tnucl:5,aa:21\n",
      "Max results per query                  \t1000\n",
      "Split database                         \t0\n",
      "Split mode                             \t2\n",
      "Split memory limit                     \t0\n",
      "Diagonal scoring                       \ttrue\n",
      "Exact k-mer matching                   \t0\n",
      "Mask residues                          \t1\n",
      "Mask lower case residues               \t0\n",
      "Minimum diagonal score                 \t15\n",
      "Spaced k-mers                          \t1\n",
      "Spaced k-mer pattern                   \t\n",
      "Local temporary path                   \t\n",
      "Rescore mode                           \t0\n",
      "Remove hits by seq. id. and coverage   \tfalse\n",
      "Sort results                           \t0\n",
      "Mask profile                           \t1\n",
      "Profile E-value threshold              \t0.1\n",
      "Global sequence weighting              \tfalse\n",
      "Allow deletions                        \tfalse\n",
      "Filter MSA                             \t1\n",
      "Maximum seq. id. threshold             \t0.9\n",
      "Minimum seq. id.                       \t0\n",
      "Minimum score per column               \t-20\n",
      "Minimum coverage                       \t0\n",
      "Select N most diverse seqs             \t1000\n",
      "Min codons in orf                      \t30\n",
      "Max codons in length                   \t32734\n",
      "Max orf gaps                           \t2147483647\n",
      "Contig start mode                      \t2\n",
      "Contig end mode                        \t2\n",
      "Orf start mode                         \t1\n",
      "Forward frames                         \t1,2,3\n",
      "Reverse frames                         \t1,2,3\n",
      "Translation table                      \t1\n",
      "Translate orf                          \t0\n",
      "Use all table starts                   \tfalse\n",
      "Offset of numeric ids                  \t0\n",
      "Create lookup                          \t0\n",
      "Add orf stop                           \tfalse\n",
      "Overlap between sequences              \t0\n",
      "Sequence split mode                    \t1\n",
      "Header split mode                      \t0\n",
      "Chain overlapping alignments           \t0\n",
      "Merge query                            \t1\n",
      "Search type                            \t0\n",
      "Search iterations                      \t1\n",
      "Start sensitivity                      \t4\n",
      "Search steps                           \t1\n",
      "Exhaustive search mode                 \tfalse\n",
      "Filter results during exhaustive search\t0\n",
      "Strand selection                       \t1\n",
      "LCA search mode                        \tfalse\n",
      "Disk space limit                       \t0\n",
      "MPI runner                             \t\n",
      "Force restart with latest tmp          \tfalse\n",
      "Remove temporary files                 \tfalse\n",
      "\n",
      "./data/mmseqs_data/result.dbtype exists already!\n",
      "./data/mmseqs_data/alignment_msa.fasta exists and will be overwritten\n",
      "result2msa ./data/mmseqs_data/protein_db ./data/mmseqs_data/database_db ./data/mmseqs_data/result ./data/mmseqs_data/alignment_msa.fasta --msa-format-mode 2 \n",
      "\n",
      "MMseqs Version:           \t13.45111\n",
      "Substitution matrix       \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Gap open cost             \tnucl:5,aa:11\n",
      "Gap extension cost        \tnucl:2,aa:1\n",
      "Allow deletions           \tfalse\n",
      "Compositional bias        \t1\n",
      "MSA format mode           \t2\n",
      "Summary prefix            \tcl\n",
      "Skip query                \tfalse\n",
      "Filter MSA                \t0\n",
      "Maximum seq. id. threshold\t0.9\n",
      "Minimum seq. id.          \t0\n",
      "Minimum score per column  \t-20\n",
      "Minimum coverage          \t0\n",
      "Select N most diverse seqs\t1000\n",
      "Threads                   \t64\n",
      "Compressed                \t0\n",
      "Verbosity                 \t3\n",
      "\n",
      "Query database size: 1 type: Aminoacid\n",
      "Target database size: 850 type: Aminoacid\n",
      "[=================================================================] 1 0s 0ms\n",
      "Time for merging to alignment_msa.fasta: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 9ms\n"
     ]
    }
   ],
   "source": [
    "temp_kl = mmseqs_cons('./data/Q72KT1.fasta','./data/850db.fasta',input_pdb = './data/Q72KT1.pdb', include_neighbors = True,method = 'rel_entr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.63694105 0.52011183 0.55386659 0.46478871 0.50182057 0.51919405\n",
      " 0.51076311 0.60423363 0.48755964 0.46478871 0.60423363 0.51465157\n",
      " 0.60264483 0.60264483 0.54737388 0.55386659 0.59181465 0.53603192\n",
      " 0.53603192 0.54737388]\n",
      "[0.55075051 0.51314026 0.49505482 0.47397989 0.48494774 0.50340174\n",
      " 0.52548613 0.55294998 0.51769536 0.51068712 0.56218254 0.53805671\n",
      " 0.55677799 0.59077585 0.55561545 0.53554546 0.53509779 0.54712757\n",
      " 0.54810723 0.55763258]\n"
     ]
    }
   ],
   "source": [
    "print(temp_spatial[0][:20])\n",
    "print(temp_spatial[1][:20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(array([  0,   1,   2,   4,   5,   6,   7,   8,  10,  11,  12,  13,  14,\n",
      "        15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  27,  29,\n",
      "        30,  35,  37,  53,  54,  56,  75,  79,  80,  81,  84, 107, 110,\n",
      "       118, 156, 169, 189, 191, 193, 196, 207, 268, 284, 290, 311, 320,\n",
      "       326, 336, 337, 338, 339, 340, 341, 343, 344, 346, 348, 353, 354,\n",
      "       355, 356, 357, 358, 359, 360, 361, 362]),)\n",
      "[3.72970145 2.27706477 2.55104645 2.17413028 2.24467686 2.22397419\n",
      " 3.14655516 3.14655516 2.20992656 3.12356565 3.12356565 2.48891467\n",
      " 2.55104645 2.97592965 2.3859667  2.3859667  2.48891467 2.88240359\n",
      " 2.97592965 3.12356565 3.14655516 2.97592965 2.16316353 2.3859667\n",
      " 2.3859667  2.55245246 2.93853774 3.14655516 3.621513   2.8605287\n",
      " 2.33840306 2.71374243 2.48891467 3.50056237 2.7019396  2.25024087\n",
      " 3.60314697 3.38719106 3.44911055 2.59805565 2.47947541 2.21116288\n",
      " 2.24021247 2.48891467 2.48891467 2.19591926 2.33122113 2.3266575\n",
      " 2.32345579 2.48891467 2.30042925 2.18232284 2.37690917 2.23606945\n",
      " 2.21500139 2.16781249 2.3491429  2.431519   2.2568749  2.2471926\n",
      " 2.4301307  2.57826422 2.83021784 2.88240359 2.48891467 2.88240359\n",
      " 2.55104645 2.3859667  3.12356565 2.97592965 3.14655516 2.88240359\n",
      " 2.48891467]\n"
     ]
    }
   ],
   "source": [
    "print(np.where(temp_spatial[0] > np.quantile(temp_spatial[0],0.8)))\n",
    "print(temp_kl[0][temp_kl[0] > np.quantile(temp_kl[0],0.8)])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
